### PR TITLE
Initial commit of sources and tests for System.Text.Encodings.CodePages

### PR DESCRIPTION
**Pull request by MattGal**

### CODE REVIEW

File: src/System.Text.Encodings.CodePages/src/System/Text/DecoderBestFitFallback.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    {
+        // Our variables
+        internal BaseCodePageEncoding encoding = null;
+        internal char[] arrayBestFit = null;
+        internal char cReplacement = '?';
+
+        internal InternalDecoderBestFitFallback(BaseCodePageEncoding encoding)
+        {
+            // Need to load our replacement characters table.
+            this.encoding = encoding;
```
Comment (joshfree) [inline src/System.Text.Encodings.CodePages/src/System/Text/DecoderBestFitFallback.cs]: was the codeformatter run on this?; curious why we still have "this.encoding" vs _encoding = encoding;

File: src/System.Text.Encodings.CodePages/src/System/Text/DecoderBestFitFallback.cs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                    Interlocked.CompareExchange<Object>(ref s_InternalSyncObject, o, null);
+                }
+                return s_InternalSyncObject;
+            }
+        }
+
+        // Constructor
+        public InternalDecoderBestFitFallbackBuffer(InternalDecoderBestFitFallback fallback)
+        {
+            this._oFallback = fallback;
```
Comment (joshfree) [inline src/System.Text.Encodings.CodePages/src/System/Text/DecoderBestFitFallback.cs]: unnecessary this

File: src/System.Text.Encodings.CodePages/src/System/Text/DecoderNLS.cs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        protected EncodingNLS m_encoding;
+        protected bool m_mustFlush;
+        internal bool m_throwOnOverflow;
+        internal int m_bytesUsed;
+        internal DecoderFallback m_fallback;
+        internal DecoderFallbackBuffer m_fallbackBuffer;
+
+        internal DecoderNLS(EncodingNLS encoding)
+        {
+            this.m_encoding = encoding;
```
Comment (joshfree) [inline src/System.Text.Encodings.CodePages/src/System/Text/DecoderNLS.cs]: another spot

**Comment by joshfree** [comment]
Looks good overall.  Just some minor feedback above where we're not following coding convention; I'd do a scrub for that.

**Comment by mellinoe** [comment]
We need to clean up the Interop code, as well, it looks like this is just the auto-generated code, which is pretty ugly and pretty confusing as well. InteropGen.cpp needs to be removed, also. It also looks like some source code was changed to use this, as Interop.mincore doesn't seem to be present, and the code is now calling directly into the Interop class.

File: src/System.Text.Encodings.CodePages/src/Interop/Interop.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
... private char _elem_210; private char _elem_211; private char _elem_212; private char _elem_213; private char _elem_214; private char _elem_215; private char _elem_216; private char _elem_217; private char _elem_218; private char _elem_219; private char _elem_220; private char _elem_221; private char _elem_222; private char _elem_223; private char _elem_224; private char _elem_225; private char _elem_226; private char _elem_227; private char _elem_228; private char _elem_229; private char _elem_230; private char _elem_231; private char _elem_232; private char _elem_233; private char _elem_234; private char _elem_235; private char _elem_236; private char _elem_237; private char _elem_238; private char _elem_239; private char _elem_240; private char _elem_241; private char _elem_242; private char _elem_243; private char _elem_244; private char _elem_245; private char _elem_246; private char _elem_247; private char _elem_248; private char _elem_249; private char _elem_250; private char _elem_251; private char _elem_252; private char _elem_253; private char _elem_254; private char _elem_255; private char _elem_256; private char _elem_257; private char _elem_258; private char _elem_259;
```
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.CodePages/src/Interop/Interop.cs]: Could we put these on separate lines?

Comment (joshfree) [inline src/System.Text.Encodings.CodePages/src/Interop/Interop.cs]: good catch :+1:

**Comment by joshfree** [comment]
Also this PR should be against the v1.0 branch and then merged up to master.

Currently it's going directly against master - this isn't correct for initial commits of libraries / tests from the closed source system.

**Pull request closed without merging**